assets/js/workflow-reducer.js

```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
 export function addModuleAction(moduleId, insertBefore) {
-  return (
-    api.addModule(getPageID(), moduleId, insertBefore)
-    .then( data => store.dispatch(changeSelectedWfModuleAction(data.id)))
-    .then( reloadWorkflowAction )
-  )
+  return function (dispatch) {
+    return (
+
+      dispatch({
+        type: ADD_MODULE,
+        payload: api.addModule(getPageID(), moduleId, insertBefore)
+      }).then( ({value, action}) => {
+
+      dispatch(
+        setSelectedWfModuleAction(value.id)
+      );
+
+    }));
+  }
 }
+registerReducerFunc(ADD_MODULE + '_FULFILLED', (state, action) => {
+  let insertBefore = action.payload.insert_before;
 
-// Call delete API, then dispatch a reload
-export function removeModuleAction(id_to_delete) {
+  delete action.payload.insert_before;
+
+  return update(state, {
+    selected_wf_module: {$set: action.payload.id},
+    workflow: {
+      wf_modules: { $splice:[ [insertBefore, 0, action.payload] ] }
+    }
+  });
+});
 
+
+// DELETE_MODULE_ACTION
+// Call delete API, then dispatch a reload
+export function deleteModuleAction(id_to_delete) {
   // If we are deleting the selected module, then set previous module in stack as selected
-  var state = store.getState();
+  let newSelectedId = null;
+  let state = store.getState();
   if (id_to_delete === state.selected_wf_module) {
 
     // Find id of previous in stack
-    var wf_modules = state.workflow.wf_modules;
-    var new_selected_id = null;
-    for (var wfm of wf_modules) {
+    let wf_modules = state.workflow.wf_modules;
+    for (let wfm of wf_modules) {
       if (wfm.id === id_to_delete)
         break;
-      new_selected_id = wfm.id;
+      newSelectedId = wfm.id;
     }
 
     // if we are deleting first module, set to new first module if any
-    if (new_selected_id === null) {
-      if (wf_modules.length > 1)
-        new_selected_id = wf_modules[1].id;
-      else
-        new_selected_id = null;
+    if (newSelectedId === null && wf_modules.length > 1) {
+      newSelectedId = wf_modules[1].id;
     }
-
-    store.dispatch(changeSelectedWfModuleAction(new_selected_id))
   }
 
-  return (
-    api.deleteModule(id_to_delete)
-    .then( reloadWorkflowAction )
-  )
-}
-
-export function wfModuleStatusAction(wfModuleID, status, error_msg='') {
-  return {
-    type : MODULE_STATUS_CHANGE,
-    id : wfModuleID,
-    status : status,
-    error_msg: error_msg
+  // If the selected module isn't changing, set it to the current selected module.
+  if (newSelectedId === null) {
+    newSelectedId = state.selected_wf_module;
   }
-}
 
-export function changeSelectedWfModuleAction(wfModuleID) {
-  return {
-    type : SELECTED_MODULE_CHANGE,
-    id : wfModuleID,
+  return function (dispatch) {
+    return (
+      // Set the new selected module before deleting to avoid errors.
+      // We do this even if the selected module isn't changing to avoid
+      // writing a subtly tricky conditional here.
```

Comment:
It does cost an extra server round-trip, however. Or maybe we do this optimistically anyway so we delete on the client without waiting?